### PR TITLE
feat: Blessings

### PR DESCRIPTION
**Pull request by kokekanon**
# Description

controller + css/html

bless

image

I know there are some visual errors, like

-  the close and history buttons are on the left. 
- the title of the mini windows is not visible  (maybe error system html? or me error syntax?)
- the images are probably out of order
- windows does not appear in the middle (maybe error system html ?   self:loadHtml() )
- verticalScroll bar in historial panel

if it was otui, I would fix it but I'm not good at css /html
image
image

 im 8.6 player. I don't know how it works. just wanted to test the new functionality, the module was easy to make 

### **Actual**

Do this and that doesn't happens

### **Expected**

Do this and that happens

## Fixes

\# (issue)

## Type of change

Please delete options that are not relevant.

  - [ ] Bug fix (non-breaking change which fixes an issue)
  - [ ] New feature (non-breaking change which adds functionality)
  - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - [ ] This change requires a documentation update

## How Has This Been Tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

  - [ ] Test A
  - [ ] Test B

**Test Configuration**:

  - Server Version:
  - Client:
  - Operating System:

## Checklist

  - [ ] My code follows the style guidelines of this project
  - [ ] I have performed a self-review of my own code
  - [ ] I checked the PR checks reports
  - [ ] I have commented my code, particularly in hard-to-understand areas
  - [ ] I have made corresponding changes to the documentation
  - [ ] My changes generate no new warnings
  - [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: modules/game_blessing/style.otui [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    anchors.left: prev.right
+    anchors.bottom : parent.bottom
+    id: name
+    width: 216
+    !text: tr('-')
+    text-align: left
+    color: #ff0000cf
+    phantom: true
+    margin-left: 5
+
```
Comment (luanluciano93) [inline modules/game_blessing/style.otui:35]: remove 2 spaces

File: modules/game_blessing/blessing.lua [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+end
+
+function setBlessing()
+    local ui = BlessingController.ui
+    ui.test:show()
+    ui.promotionStatus2:show()
+    ui.promotionStatus:show()
+    ui.blessingHistory:hide()
+    ui.historyButton:setText("History")
+end
```
Comment (luanluciano93) [inline modules/game_blessing/blessing.lua:54]: space

File: modules/game_blessing/blessing.html [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  .contenedor {
+    width: 580;
+    display: block;
+  }
+
+  .record-blessing {
+    width: 580;
+    height: 120;
+    display: block;
+  }
```
Comment (luanluciano93) [inline modules/game_blessing/blessing.html:48]: space

File: modules/game_blessing/blessing.lua [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+            "%, #f75f5f} chance that you will lose your equipped container on your next\ndeath.")
+
+    BlessingController.ui.promotionStatus.equipmentLoss:setColoredText(
+        "- There is a {" .. data.equipPveLoss .. "%, #f75f5f} chance that you will lose items upon your next death.")
+
+    BlessingController.ui.blessingHistory.data:destroyChildren()
+
+    for index, entry in ipairs(data.logs) do
+        local row = g_ui.createWidget("historyData", BlessingController.ui.blessingHistory.data)
+        local date = os.date("%Y-%m-%d, %H:%M:%S", entry.timestam)
```
Comment (luanluciano93) [inline modules/game_blessing/blessing.lua]: timestamp

File: src/client/game.h [hunk 5]
```diff
@@ -394,7 +424,7 @@ class Game
                             const std::vector<std::tuple<uint32_t, std::string, std::string, uint8_t, std::string, uint16_t, uint8_t, uint64_t>>& highscores, uint32_t entriesTs);
 
     void imbuementDurations(bool isOpen = false);
-
+    void requestBless();
```
Comment (luanluciano93) [inline src/client/game.h:516]: space

File: src/client/protocolgameparse.cpp [hunk 6]
```diff
@@ -3474,36 +3474,37 @@ void ProtocolGame::parseGameNews(const InputMessagePtr& msg)
...
-        msg->getU32(); // timestamp
-        msg->getU8(); // color message (0 = white loss, 1 = red)
-        msg->getString(); // history message
-    }
-
-    // TODO: implement bless dialog usage
+    BlessDialogData data;
+    data.totalBless = msg->getU8();
+
+    for (uint8_t i = 0; i < data.totalBless; ++i) {
```
Comment (luanluciano93) [inline src/client/protocolgameparse.cpp]: I don't think it's necessary to store totalBless in data

**Review by luanluciano93 — CHANGES_REQUESTED** [review]

File: src/client/protocolgameparse.cpp [hunk 7]
```diff
@@ -3474,36 +3474,37 @@ void ProtocolGame::parseGameNews(const InputMessagePtr& msg)
...
+    data.pvpMinXpLoss = msg->getU8();
+    data.pvpMaxXpLoss = msg->getU8();
+    data.pveExpLoss = msg->getU8();
+    data.equipPvpLoss = msg->getU8();
+    data.equipPveLoss = msg->getU8();
+    data.skull = msg->getU8();
+    data.aol = msg->getU8();
+
+    uint8_t logCount = msg->getU8();
+    for (uint8_t i = 0; i < logCount; ++i) {
```
Comment (luanluciano93) [inline src/client/protocolgameparse.cpp]:
```suggestion
    const uint8_t logCount = msg->getU8();
    for (auto i = 0; i < logCount; ++i) {
```

File: src/client/protocolgameparse.cpp (hunk 6 above)
Comment (luanluciano93) [inline src/client/protocolgameparse.cpp]:
```suggestion
    for (auto i = 0; i < data.totalBless; ++i) {
```

File: modules/game_blessing/blessing.lua [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    row2.name:setColor("#c0c0c0")
+
+    for index, entry in ipairs(data.logs) do
+        local row = g_ui.createWidget("historyData", BlessingController.ui.blessingHistory:getChildByIndex(1))
+        local date = os.date("%Y-%m-%d, %H:%M:%S", entry.timestamp)
+        row:setBackgroundColor(index % 2 == 0 and "#ffffff12" or "#00000012")
+        row.rank:setText(date)
+        row.name:setText(entry.historyMessage)
+    end
+
```
Comment (luanluciano93) [inline modules/game_blessing/blessing.lua]: remove space

File: modules/game_blessing/blessing.lua [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    end
+    g_game.requestBless()
+    BlessingController.ui:show()
+    BlessingController.ui:raise()
+    BlessingController.ui:focus()
+    setBlessing()
+end
+
+function onUpdateBlessDialog(data)
+
```
Comment (luanluciano93) [inline modules/game_blessing/blessing.lua]: remove space

**Review by luanluciano93 — CHANGES_REQUESTED** [review]

**Comment by kokekanon** [comment]
need to make a `git reset --hard hash` sorry open other pr

**Pull request closed without merging**